### ISSUE TITLE
evaluator: disjunctions don't have distinct values

### ISSUE THREAD

**Issue by uhthomas** [cue-lang/cue#2421]
### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
0.5.0
</pre>

Also happens when built from [HEAD](https://github.com/cue-lang/cue/commit/251790edc0f6b73fe1dcd4fb41d408b0d962c179).

### Does this issue reproduce with the latest stable release?

Yes

### What did you do?

```cue
#Parent & {
    config: {
        name: "some name"
        value: #B
    }
}

#Parent: {
    config: #Config
    let _config = config
    out: #Z & {config: _config}
}

#Config: {
    name: string | *"unset"
    value: #A | #B | #C
}

#A: {}

#B: {}

#C: {}

#Z: {
    config: #Config
    name: config.name
    isA: (config.value & #A) != _|_
    isB: (config.value & #B) != _|_
    isC: (config.value & #C) != _|_
}
```

### What did you expect to see?

```json
{
    "config": {
        "name": "some name",
        "value": {}
    },
    "out": {
        "config": {
            "name": "some name",
            "value": {}
        },
        "name": "some name",
        "isA": false,
        "isB": true,
        "isC": false
    }
}
```

### What did you see instead?

```json
{
    "config": {
        "name": "some name",
        "value": {}
    },
    "out": {
        "config": {
            "name": "some name",
            "value": {}
        },
        "name": "some name",
        "isA": true,
        "isB": true,
        "isC": true
    }
}
```

**Comment by mvdan** [comment]
Still seems to happen as of 049d6e12211e8331185b836acf93ee5efc65addd.

**Comment by mvdan** [comment]
@uhthomas I realize it's been years, but could you please explain what is your reasoning behind why you think the results should be "false, true, false"?

As far as I can tell, `#A, #B, #C` are identical struct definitions, so they can be used interchangeably. If you really want them to be treated as distinct types, you can add a hidden discriminator field to them, for example:

```
exec cue-v0.15.0 export in.cue

exec cue export in.cue

-- in.cue --
#Parent & {
    config: {
        name: "some name"
        value: #B
    }
}

#Parent: {
    config: #Config
    let _config = config
    out: #Z & {config: _config}
}

#Config: {
    name: string | *"unset"
    value: #A | #B | #C
}

#A: {_kind: "A"}
#B: {_kind: "B"}
#C: {_kind: "C"}

#Z: {
    config: #Config
    name: config.name
    isA: (config.value & #A) != _|_
    isB: (config.value & #B) != _|_
    isC: (config.value & #C) != _|_
}
```
```
> exec cue export in.cue
[stdout]
{
    "config": {
        "name": "some name",
        "value": {}
    },
    "out": {
        "config": {
            "name": "some name",
            "value": {}
        },
        "name": "some name",
        "isA": false,
        "isB": true,
        "isC": false
    }
}
```

**Comment by uhthomas** [comment]
Good morning! I guess intuitively I had anticipated it may work more like Go, where types are unique.

```go
package main

import "fmt"

type A struct{}

type B struct{}

func main() {
	var out any = A{}
	_, isA := out.(A)
	_, isB := out.(B)
	fmt.Println(isA, isB)
}
// true false
```

**Comment by mvdan** [comment]
Ah, indeed CUE doesn't have "named types" which are unique in that way :) I'm not sure how to better document that, given that it's a preconception that one might bring from Go, but not necessarily from other languages. Any ideas?

**Comment by uhthomas** [comment]
Maybe it's best to document a solution to what I was trying to achieve? I think at the time I had a list of Kubernetes objects, and I wanted to check what type they were. `(object & #ConfigMap) != _|_` doesn't work for that, so the only way to do it is by checking the `kind` field such `object.apiVersion == "v1" && object.kind == "ConfigMap"`.

**Comment by mvdan** [comment]
Yes, I was thinking along the lines of a discriminator fields doc page to outline general patterns and solutions in that space. I've raised https://github.com/cue-lang/cue/issues/4259 for this purpose.

> I had a list of Kubernetes objects, and I wanted to check what type they were. `(object & #ConfigMap) != _|_` doesn't work for that

Hm, this should work, because of the `kind` field. That's basically what my example above does with `_kind`.

**Comment by uhthomas** [comment]
It's possible that this is just because of the cue generated from `get go`. The api version and kind fields are empty, whereas the new modules on the registry are pre-filled. 

https://registry.cue.works/source/cue.dev/x/k8s.io/api/core/v1/schema.cue#L549

**Comment by mvdan** [comment]
Ah, that makes sense. I guess `cue get go` will always have this limitation because Go types don't carry constant values, but JSON Schema can do that.

If you can think of a way to extend `cue get go` to do this, much like it was recently extended to support k8s optional and required annotations, then please raise an issue or a PR. Otherwise, it's probably best to stick to the JSONSchema-derived definitions.

I'll close this in favor of https://github.com/cue-lang/cue/issues/4259 for the docs, then. Thanks!
